File: services/core/src/cortex_core/files.py
"""Private immutable bytes with lease-based parsing and permission-aware retrieval."""

import base64
import binascii
import hashlib
import json
import os
import subprocess
import sys
from uuid import uuid4

from .auth import CoreError
from .contracts import SourceInput
from .service import digest, encoded, one, run


def parse_bytes(raw, filename):
    extension = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""
    try:
        result = subprocess.run(
            [sys.executable, "-m", "cortex_core.parse_document", extension],
            input=raw,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            timeout=15,
            env={"PATH": os.defpath, "PYTHONNOUSERSITE": "1"},
            check=False,
        )
    except subprocess.TimeoutExpired:
        return {"error_code": "PARSE_TIMEOUT"}
    if result.returncode != 0:
        return {"error_code": "PARSE_RESOURCE_FAILURE"}
    try:
        return json.loads(result.stdout)
    except (ValueError, UnicodeDecodeError):
        return {"error_code": "PARSE_FAILED"}


class FileService:
    def __init__(self, corpus):
        self.corpus, self.k, self.db = corpus, corpus.knowledge, corpus.db

    def _file(self, conn, p, domain, ident):
        row = one(
            conn,
            """SELECT f.* FROM cf_files f JOIN cf_collections c
            ON c.tenant_id=f.tenant_id AND c.domain_id=f.domain_id AND c.id=f.collection_id
            WHERE f.tenant_id=:tenant AND f.domain_id=:domain AND f.id=:id
            AND f.allowed_subjects @> CAST(:acl AS jsonb) AND c.allowed_subjects @> CAST(:acl AS jsonb)""",
            **self.k.keys(p, domain),
            id=ident,
            acl=encoded([p.subject]),
        )
        if not row:
            raise CoreError("NOT_FOUND", "File not found", 404)
        if row["source_id"]:
            self.k._source(conn, p, domain, row["source_id"])
        return row

    @staticmethod
    def view(row):
        return {
            **{
                k: row[k]
                for k in (
                    "id",
                    "collection_id",
                    "filename",
                    "content_hash",
                    "status",
                    "source_id",
                    "error_code",
                    "attempts",
                    "spans",
                )
            },
            "size_bytes": len(row["data"]),
        }

    def upload(self, p, domain, collection, data):
        try:
            raw = base64.b64decode(data.content_base64, validate=True)
        except (ValueError, binascii.Error) as exc:
            raise CoreError("VALIDATION_FAILED", "Invalid base64", 422) from exc
        if not raw or len(raw) > 500000:
            raise CoreError("DOCUMENT_LIMIT", "File must contain 1 to 500,000 bytes", 422)
        with self.db.transaction(p, domain, corpus=True) as conn:
            self.k._domain(conn, p, domain, lock=True)
            col = self.corpus._collection(conn, p, domain, collection)
            self.k._validate_subjects(conn, p, domain, data.allowed_subjects)
            if p.subject not in data.allowed_subjects or not set(data.allowed_subjects) <= set(
                col["allowed_subjects"]
            ):
                raise CoreError(
                    "VALIDATION_FAILED",
                    "File readers must include uploader and be within collection readers",
                    422,
                )
            h = hashlib.sha256(raw).hexdigest()
            fingerprint = digest(
                {
                    "collection": collection,
                    "filename": data.filename,
                    "hash": h,
                    "readers": sorted(set(data.allowed_subjects)),
                }
            )
            old = one(
                conn,
                """SELECT * FROM cf_files WHERE tenant_id=:tenant AND domain_id=:domain
                AND author=:author AND idempotency_key=:key""",
                **self.k.keys(p, domain),
                author=p.subject,
                key=data.idempotency_key,
            )
            if old:
                self._file(conn, p, domain, old["id"])
                if old["request_hash"] != fingerprint:
                    raise CoreError("IDEMPOTENCY_CONFLICT", "Upload key reused")
                return self.view(old)
            ident = str(uuid4())
            run(
                conn,
                """INSERT INTO cf_files(tenant_id,domain_id,id,collection_id,author,filename,
                content_hash,data,allowed_subjects,idempotency_key,request_hash)
                VALUES(:tenant,:domain,:id,:collection,:author,:filename,:hash,:data,CAST(:acl AS jsonb),:key,:fingerprint)""",
                **self.k.keys(p, domain),
                id=ident,
                collection=collection,
                author=p.subject,
                filename=data.filename,
                hash=h,
                data=raw,
                acl=encoded(sorted(set(data.allowed_subjects))),
                key=data.idempotency_key,
                fingerprint=fingerprint,
            )
            return self.view(self._file(conn, p, domain, ident))

    def detail(self, p, domain, ident):
        with self.db.transaction(p, domain) as conn:
            return self.view(self._file(conn, p, domain, ident))

    def download(self, p, domain, ident):
        with self.db.transaction(p, domain) as conn:
            return bytes(self._file(conn, p, domain, ident)["data"])

    def listing(self, p, domain, limit, after, pending):
        from .workspace import WorkspaceService

        with self.db.transaction(p, domain) as conn:
            return WorkspaceService(self.k)._page(
                conn,
                p,
                domain,
                "cf_files",
                limit,
                after,
                lambda row: self._file(conn, p, domain, row["id"]),
                self.view,
                "AND (status='pending' OR (status='processing' AND lease_until<now()))"
                if pending
                else "",
            )

    def process(self, p, domain, ident):
        with self.db.transaction(p, domain, corpus=True) as conn:
            self.k._domain(conn, p, domain, lock=True)
            row = self._file(conn, p, domain, ident)
            if row["status"] in ("succeeded", "failed"):
                return self.view(row)
            if row["status"] == "cancelled":
                raise CoreError("FILE_CANCELLED", "File processing was cancelled")
            if (
                row["status"] == "processing"
                and one(conn, "SELECT :until>now() AS active", until=row["lease_until"])["active"]
            ):
                raise CoreError("FILE_BUSY", "Another worker holds the parsing lease")
            lease = str(uuid4())
            run(
                conn,
                """UPDATE cf_files SET status='processing',attempts=attempts+1,
                lease_token=:lease,lease_until=now()+interval '60 seconds' WHERE tenant_id=:tenant AND domain_id=:domain AND id=:id""",
                **self.k.keys(p, domain),
                id=ident,
                lease=lease,
            )
            raw, filename = bytes(row["data"]), row["filename"]
        result = parse_bytes(raw, filename)
        with self.db.transaction(p, domain, corpus=True) as conn:
            self.k._domain(conn, p, domain, lock=True)
            row = self._file(conn, p, domain, ident)
            if row["status"] != "processing" or row["lease_token"] != lease:
                raise CoreError("STALE_LEASE", "Parsing result no longer belongs to the active job")
            source_id, error = None, result.get("error_code")
            if not error:
                try:
                    with conn.begin_nested():
                        source = self.k._create_source(
                            conn,
                            p,
                            domain,
                            SourceInput(
                                title=filename,
                                location=f"document://{row['collection_id']}/{filename}/{row['content_hash']}",
                                content=result["content"],
                                allowed_subjects=row["allowed_subjects"],
                            ),
                        )
                        source_id = source["id"]
                        run(
                            conn,
                            """INSERT INTO cf_collection_sources VALUES(:tenant,:domain,:collection,:source)
                            ON CONFLICT DO NOTHING""",
                            **self.k.keys(p, domain),
                            collection=row["collection_id"],
                            source=source_id,
                        )
                except CoreError as exc:
                    error = exc.code
            run(
                conn,
                """UPDATE cf_files SET status=:status,source_id=:source,error_code=:error,
                spans=CAST(:spans AS jsonb),lease_token=NULL,lease_until=NULL
                WHERE tenant_id=:tenant AND domain_id=:domain AND id=:id""",
                **self.k.keys(p, domain),
                id=ident,
                status="failed" if error else "succeeded",
                source=source_id,
                error=error,
                spans=encoded(result.get("spans", []) if not error else []),
            )
            return self.view(self._file(conn, p, domain, ident))

    def control(self, p, domain, ident, action):
        with self.db.transaction(p, domain, corpus=True) as conn:
            self.k._domain(conn, p, domain, lock=True)
            row = self._file(conn, p, domain, ident)
            target = "pending" if action == "retry" else "cancelled"
            allowed = ("failed",) if action == "retry" else ("pending", "processing")
            if row["status"] in allowed:
                run(
                    conn,
                    """UPDATE cf_files SET status=:status,error_code=NULL,lease_token=NULL,lease_until=NULL
                    WHERE tenant_id=:tenant AND domain_id=:domain AND id=:id""",
                    **self.k.keys(p, domain),
                    id=ident,
                    status=target,
                )
            return self.view(self._file(conn, p, domain, ident))

File: services/rust-core/src/files.rs
//! Immutable file bytes and current corpus ACLs. Parsing remains an explicit separate operation.
use crate::{auth::Principal, error::CoreError, retrieval, server::StateData};
use axum::{
    Json, Router,
    extract::{Path, Query, State},
    response::{IntoResponse, Response},
    routing::{get, post},
};
use base64::{
    Engine,
    engine::{DecodePaddingMode, GeneralPurpose, GeneralPurposeConfig},
};
use http::{HeaderMap, StatusCode};
use serde::Deserialize;
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use sqlx::{Acquire, Postgres, Row, Transaction, postgres::PgRow};
use std::collections::BTreeSet;
use uuid::Uuid;
fn invalid() -> CoreError {
    CoreError {
        code: "VALIDATION_FAILED",
        message: "Invalid file arguments or readers",
        status: StatusCode::UNPROCESSABLE_ENTITY,
    }
}
fn uuid(s: &str) -> Result<String, CoreError> {
    Uuid::parse_str(s)
        .map(|x| x.to_string())
        .map_err(|_| CoreError::invalid_uuid())
}
fn twenty() -> usize {
    20
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Input {
    filename: String,
    content_base64: String,
    allowed_subjects: Vec<String>,
    idempotency_key: String,
}
#[derive(Deserialize)]
struct Page {
    #[serde(default = "twenty")]
    limit: usize,
    after: Option<Uuid>,
    #[serde(default)]
    pending: bool,
}
const ACCESS: &str = "SELECT f.* FROM cf_files f JOIN cf_collections c ON c.tenant_id=f.tenant_id AND c.domain_id=f.domain_id AND c.id=f.collection_id LEFT JOIN cf_sources s ON s.tenant_id=f.tenant_id AND s.domain_id=f.domain_id AND s.id=f.source_id WHERE f.tenant_id=$1 AND f.domain_id=$2 AND f.allowed_subjects ? $3 AND c.allowed_subjects ? $3 AND (f.source_id IS NULL OR s.allowed_subjects ? $3)";
async fn file(
    tx: &mut Transaction<'_, Postgres>,
    p: &Principal,
    d: &str,
    id: &str,
) -> Result<PgRow, CoreError> {
    sqlx::query(&format!("{ACCESS} AND f.id=$4"))
        .bind(&p.tenant)
        .bind(d)
        .bind(&p.subject)
        .bind(id)
        .fetch_optional(&mut **tx)
        .await
        .map_err(CoreError::sql)?
        .ok_or_else(CoreError::not_found)
}
fn view(r: &PgRow) -> Value {
    json!({"id":r.get::<String,_>("id"),"collection_id":r.get::<String,_>("collection_id"),"filename":r.get::<String,_>("filename"),"content_hash":r.get::<String,_>("content_hash"),"status":r.get::<String,_>("status"),"source_id":r.get::<Option<String>,_>("source_id"),"error_code":r.get::<Option<String>,_>("error_code"),"attempts":r.get::<i32,_>("attempts"),"spans":r.get::<Value,_>("spans"),"size_bytes":r.get::<Vec<u8>,_>("data").len()})
}
async fn write_tx<'a>(
    s: &'a StateData,
    p: &Principal,
    d: &str,
) -> Result<Transaction<'a, Postgres>, CoreError> {
    s.db.locked_permission_transaction(
        p,
        d,
        &["owner", "corpus_manager"],
        "Corpus management permission required",
    )
    .await
}
fn decode(s: &str) -> Result<Vec<u8>, CoreError> {
    let bare = s.trim_end_matches('=');
    if bare.is_empty() {
        return Err(invalid());
    }
    let padding = s.len() - bare.len();
    match bare.len() % 4 {
        0 if padding == 0 => {}
        2 if padding == 2 => {}
        3 if padding == 1 => {}
        _ => return Err(invalid()),
    }
    GeneralPurpose::new(
        &base64::alphabet::STANDARD,
        GeneralPurposeConfig::new()
            .with_decode_padding_mode(DecodePaddingMode::Indifferent)
            .with_decode_allow_trailing_bits(true),
    )
    .decode(bare)
    .map_err(|_| invalid())
}
pub fn routes() -> Router<StateData> {
    Router::new()
        .route(
            "/v1/domains/{domain}/collections/{collection}/files",
            post(upload),
        )
        .route("/v1/domains/{domain}/files", get(list))
        .route("/v1/domains/{domain}/files/{ident}", get(read))
        .route("/v1/domains/{domain}/files/{ident}/download", get(download))
        .route("/v1/domains/{domain}/files/{ident}/retry", post(retry))
        .route("/v1/domains/{domain}/files/{ident}/process", post(process))
        .route("/v1/domains/{domain}/files/{ident}/cancel", post(cancel))
}
async fn upload(
    State(s): State<StateData>,
    Path((d, c)): Path<(String, String)>,
    h: HeaderMap,
    input: Result<Json<Input>, axum::extract::rejection::JsonRejection>,
) -> Result<(StatusCode, Json<Value>), CoreError> {
    let p = s.auth.authenticate(&h)?;
    let d = uuid(&d)?;
    let c = uuid(&c)?;
    let Json(i) = input.map_err(|_| invalid())?;
    if !(1..=200).contains(&i.filename.chars().count())
        || i.filename
            .chars()
            .any(|c| c == '/' || c == '\\' || (c as u32) < 32)
        || !(1..=666668).contains(&i.content_base64.len())
        || !(1..=100).contains(&i.allowed_subjects.len())
        || !(8..=128).contains(&i.idempotency_key.chars().count())
    {
        return Err(invalid());
    }
    let raw = decode(&i.content_base64)?;
    if raw.is_empty() || raw.len() > 500000 {
        return Err(CoreError {
            code: "DOCUMENT_LIMIT",
            message: "File must contain 1 to 500000 bytes",
            status: StatusCode::UNPROCESSABLE_ENTITY,
        });
    }
    let mut tx = write_tx(&s, &p, &d).await?;
    let col:Value=sqlx::query_scalar("SELECT allowed_subjects FROM cf_collections WHERE tenant_id=$1 AND domain_id=$2 AND id=$3 AND allowed_subjects ? $4").bind(&p.tenant).bind(&d).bind(&c).bind(&p.subject).fetch_optional(&mut *tx).await.map_err(CoreError::sql)?.ok_or_else(CoreError::not_found)?;
    let members: Vec<String> = sqlx::query_scalar(
        "SELECT subject FROM cf_memberships WHERE tenant_id=$1 AND domain_id=$2 FOR SHARE",
    )
    .bind(&p.tenant)
    .bind(&d)
    .fetch_all(&mut *tx)
    .await
    .map_err(CoreError::sql)?;
    let acl: BTreeSet<String> = i.allowed_subjects.into_iter().collect();
    let col: Vec<String> = serde_json::from_value(col).map_err(|_| CoreError::database())?;
    if !acl.contains(&p.subject) || !acl.iter().all(|x| members.contains(x) && col.contains(x)) {
        return Err(invalid());
    }
    let hash = format!("{:x}", Sha256::digest(&raw));
    let fingerprint = crate::canonical::digest(
        &json!({"collection":c,"filename":i.filename,"hash":hash,"readers":acl}),
    )
    .map_err(|_| invalid())?;
    let old=sqlx::query("SELECT id,request_hash FROM cf_files WHERE tenant_id=$1 AND domain_id=$2 AND author=$3 AND idempotency_key=$4").bind(&p.tenant).bind(&d).bind(&p.subject).bind(&i.idempotency_key).fetch_optional(&mut *tx).await.map_err(CoreError::sql)?;
    let result = if let Some(old) = old {
        let r = file(&mut tx, &p, &d, &old.get::<String, _>("id")).await?;
        if old.get::<String, _>("request_hash") != fingerprint {
            return Err(CoreError {
                code: "IDEMPOTENCY_CONFLICT",
                message: "Upload key reused",
                status: StatusCode::CONFLICT,
            });
        }
        view(&r)
    } else {
        let r=sqlx::query("INSERT INTO cf_files(tenant_id,domain_id,id,collection_id,author,filename,content_hash,data,allowed_subjects,idempotency_key,request_hash) VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11) RETURNING *").bind(&p.tenant).bind(&d).bind(Uuid::new_v4().to_string()).bind(&c).bind(&p.subject).bind(&i.filename).bind(&hash).bind(&raw).bind(json!(acl)).bind(&i.idempotency_key).bind(&fingerprint).fetch_one(&mut *tx).await.map_err(CoreError::sql)?;
        view(&r)
    };
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    Ok((StatusCode::ACCEPTED, Json(result)))
}
async fn read(
    State(s): State<StateData>,
    Path((d, id)): Path<(String, String)>,
    h: HeaderMap,
) -> Result<Json<Value>, CoreError> {
    let p = s.auth.authenticate(&h)?;
    let d = uuid(&d)?;
    let id = uuid(&id)?;
    let mut tx = retrieval::transaction(&s, &p, &d).await?;
    let result = view(&file(&mut tx, &p, &d, &id).await?);
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    Ok(Json(result))
}
async fn download(
    State(s): State<StateData>,
    Path((d, id)): Path<(String, String)>,
    h: HeaderMap,
) -> Result<Response, CoreError> {
    let p = s.auth.authenticate(&h)?;
    let d = uuid(&d)?;
    let id = uuid(&id)?;
    let mut tx = retrieval::transaction(&s, &p, &d).await?;
    let raw: Vec<u8> = file(&mut tx, &p, &d, &id).await?.get("data");
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    Ok((
        [
            ("content-type", "application/octet-stream"),
            ("content-disposition", "attachment; filename=\"source.bin\""),
            ("x-content-type-options", "nosniff"),
            ("cache-control", "no-store"),
        ],
        raw,
    )
        .into_response())
}
async fn list(
    State(s): State<StateData>,
    Path(d): Path<String>,
    h: HeaderMap,
    input: Result<Query<Page>, axum::extract::rejection::QueryRejection>,
) -> Result<Json<Value>, CoreError> {
    let p = s.auth.authenticate(&h)?;
    let d = uuid(&d)?;
    let Query(q) = input.map_err(|_| invalid())?;
    if !(1..=100).contains(&q.limit) {
        return Err(invalid());
    }
    let mut tx = retrieval::transaction(&s, &p, &d).await?;
    let rows=sqlx::query(&format!("{ACCESS} AND f.id>$4 AND (NOT $5 OR f.status='pending' OR (f.status='processing' AND f.lease_until<now())) ORDER BY f.id LIMIT $6")).bind(&p.tenant).bind(&d).bind(&p.subject).bind(q.after.map(|x|x.to_string()).unwrap_or_default()).bind(q.pending).bind((q.limit+1)as i64).fetch_all(&mut *tx).await.map_err(CoreError::sql)?;
    let next = if rows.len() > q.limit {
        json!(rows[q.limit - 1].get::<String, _>("id"))
    } else {
        Value::Null
    };
    let result =
        json!({"items":rows.iter().take(q.limit).map(view).collect::<Vec<_>>(),"next_after":next});
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    Ok(Json(result))
}
async fn retry(
    State(s): State<StateData>,
    Path((d, id)): Path<(String, String)>,
    h: HeaderMap,
) -> Result<Json<Value>, CoreError> {
    control(s, d, id, h, true).await
}
async fn cancel(
    State(s): State<StateData>,
    Path((d, id)): Path<(String, String)>,
    h: HeaderMap,
) -> Result<Json<Value>, CoreError> {
    control(s, d, id, h, false).await
}
async fn control(
    s: StateData,
    d: String,
    id: String,
    h: HeaderMap,
    retry: bool,
) -> Result<Json<Value>, CoreError> {
    let p = s.auth.authenticate(&h)?;
    let d = uuid(&d)?;
    let id = uuid(&id)?;
    let mut tx = write_tx(&s, &p, &d).await?;
    let r = file(&mut tx, &p, &d, &id).await?;
    let status: String = r.get("status");
    if (retry && status == "failed") || (!retry && (status == "pending" || status == "processing"))
    {
        sqlx::query("UPDATE cf_files SET status=$4,error_code=NULL,lease_token=NULL,lease_until=NULL WHERE tenant_id=$1 AND domain_id=$2 AND id=$3").bind(&p.tenant).bind(&d).bind(&id).bind(if retry{"pending"}else{"cancelled"}).execute(&mut *tx).await.map_err(CoreError::sql)?;
    }
    let result = view(&file(&mut tx, &p, &d, &id).await?);
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    Ok(Json(result))
}
static PARSE_SLOTS: tokio::sync::Semaphore = tokio::sync::Semaphore::const_new(2);

async fn process(
    State(s): State<StateData>,
    Path((d, id)): Path<(String, String)>,
    h: HeaderMap,
) -> Result<Json<Value>, CoreError> {
    let p = s.auth.authenticate(&h)?;
    let d = uuid(&d)?;
    let id = uuid(&id)?;
    let mut tx = write_tx(&s, &p, &d).await?;
    let r = file(&mut tx, &p, &d, &id).await?;
    let status: String = r.get("status");
    if status == "succeeded" || status == "failed" {
        let result = view(&r);
        p.check_fresh()?;
        tx.commit().await.map_err(CoreError::sql)?;
        return Ok(Json(result));
    };
    if status == "cancelled" {
        return Err(CoreError {
            code: "FILE_CANCELLED",
            message: "File processing cancelled",
            status: StatusCode::CONFLICT,
        });
    }
    let active:bool=sqlx::query_scalar("SELECT COALESCE(lease_until>now(),false) FROM cf_files WHERE tenant_id=$1 AND domain_id=$2 AND id=$3").bind(&p.tenant).bind(&d).bind(&id).fetch_one(&mut *tx).await.map_err(CoreError::sql)?;
    if status == "processing" && active {
        return Err(CoreError {
            code: "FILE_BUSY",
            message: "Parsing lease already held",
            status: StatusCode::CONFLICT,
        });
    }
    let _slot = PARSE_SLOTS.try_acquire().map_err(|_| CoreError {
        code: "FILE_BUSY",
        message: "Native parser capacity is occupied",
        status: StatusCode::SERVICE_UNAVAILABLE,
    })?;
    let attempt = r
        .get::<i32, _>("attempts")
        .checked_add(1)
        .ok_or_else(CoreError::database)?;
    let lease = Uuid::new_v4().to_string();
    sqlx::query("UPDATE cf_files SET status='processing',attempts=$4,lease_token=$5,lease_until=now()+interval '60 seconds' WHERE tenant_id=$1 AND domain_id=$2 AND id=$3").bind(&p.tenant).bind(&d).bind(&id).bind(attempt).bind(&lease).execute(&mut *tx).await.map_err(CoreError::sql)?;
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    let result =
        crate::document_parser::parse(r.get("data"), &r.get::<String, _>("filename")).await;
    let mut tx = write_tx(&s, &p, &d).await?;
    let r = file(&mut tx, &p, &d, &id).await?;
    if r.get::<String, _>("status") != "processing"
        || r.get::<Option<String>, _>("lease_token").as_deref() != Some(&lease)
    {
        return Err(CoreError {
            code: "STALE_LEASE",
            message: "Parsing result belongs to an obsolete job",
            status: StatusCode::CONFLICT,
        });
    }
    let mut error = result.error_code;
    let mut source = None;
    if error.is_none() {
        let mut nested = tx.begin().await.map_err(CoreError::sql)?;
        let outcome:Result<String,CoreError>=async{let source=crate::sources::create_in_transaction(&mut nested,&p,&d,crate::sources::SourceInput{title:r.get("filename"),location:format!("document://{}/{}/{}",r.get::<String,_>("collection_id"),r.get::<String,_>("filename"),r.get::<String,_>("content_hash")),content:result.content.ok_or_else(CoreError::database)?,allowed_subjects:serde_json::from_value(r.get("allowed_subjects")).map_err(|_|CoreError::database())?,supersedes:None}).await?;let source=source["id"].as_str().ok_or_else(CoreError::database)?.to_owned();sqlx::query("INSERT INTO cf_collection_sources(tenant_id,domain_id,collection_id,source_id) VALUES($1,$2,$3,$4) ON CONFLICT DO NOTHING").bind(&p.tenant).bind(&d).bind(r.get::<String,_>("collection_id")).bind(&source).execute(&mut *nested).await.map_err(CoreError::sql)?;Ok(source)}.await;
        match outcome {
            Ok(id) => {
                nested.commit().await.map_err(CoreError::sql)?;
                source = Some(id)
            }
            Err(e) => {
                nested.rollback().await.map_err(CoreError::sql)?;
                if e.status.is_server_error()
                    || e.code == "CONCURRENT_CHANGE"
                    || e.status == StatusCode::UNAUTHORIZED
                {
                    return Err(e);
                }
                error = Some(e.code.to_owned())
            }
        }
    }
    sqlx::query("UPDATE cf_files SET status=$4,source_id=$5,error_code=$6,spans=$7,lease_token=NULL,lease_until=NULL WHERE tenant_id=$1 AND domain_id=$2 AND id=$3").bind(&p.tenant).bind(&d).bind(&id).bind(if error.is_some(){"failed"}else{"succeeded"}).bind(source).bind(&error).bind(if error.is_none(){json!(result.spans)}else{json!([])}).execute(&mut *tx).await.map_err(CoreError::sql)?;
    let result = view(&file(&mut tx, &p, &d, &id).await?);
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    Ok(Json(result))
}
